src/ioabc.cpp

```diff
@@ -304,44 +305,100 @@ void ABCInput::CalcUnitNoteLength()
+        LayerElement *element = NULL;
+        if (m_containerElement.m_element && (ElementType::Tuplet == m_containerElement.m_type)) {
+            element = m_containerElement.m_element;
+            element->AddChild(beam);
         }
+        // otherwise default to it being beam
         else {
-            for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
-                m_layer->AddChild(*iter);
-            }
+            element = beam;
+        }
+        m_layer->AddChild(element);
+    }
+    else {
+        for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
+            m_layer->AddChild(*iter);
         }
     }
+    m_containerElement = {};
     m_noteStack.clear();
 }
 
-void ABCInput::AddTuplet()
+int ABCInput::ParseTuplet(const std::string &musicCode, int index)
 {
-    if (!m_noteStack.size()) {
-        return;
+    constexpr std::string_view tupletElements = "(:0123456789 ";
+    const size_t tupletEnd = musicCode.find_first_not_of(tupletElements, ++index);
+    const std::string tupletStr = musicCode.substr(index, tupletEnd - index);
+
+    Tuplet *tuplet = new Tuplet();
+    size_t separator = tupletStr.find_first_of(":");
+    // Get tuplet number first 9:_:_
+    int tupletNum = 0;
+    if (separator != std::string::npos) {
+        const std::string tupletNumStr = tupletStr.substr(0, separator);
+        tupletNum = atoi(tupletNumStr.data());
+        ++separator;
     }
     else {
-        Tuplet *tuplet = new Tuplet();
-        for (auto iter = m_noteStack.begin(); iter != m_noteStack.end(); ++iter) {
-            tuplet->AddChild(*iter);
+        tupletNum = atoi(tupletStr.c_str());
+    }
+    // Get tuplet number base _:3:_
+    int tupletNumbase = 0;
+    if (separator != std::string::npos) {
+        size_t secondSeparator = tupletStr.find_first_of(":", separator);
+        if (secondSeparator != std::string::npos) {
+            if (secondSeparator != separator) {
+                std::string tupletNumbaseStr = tupletStr.substr(separator, secondSeparator - separator);
+                tupletNumbase = atoi(tupletNumbaseStr.data());
+                separator = secondSeparator + 1;
+            }
+        }
+        else {
+            std::string tupletNumbaseStr = tupletStr.substr(separator);
+            tupletNumbase = atoi(tupletNumbaseStr.data());
+            separator = secondSeparator + 1;
         }
-        m_layer->AddChild(tuplet);
     }
-    m_noteStack.clear();
+    // List of tuplets with default base of 3
+    const std::unordered_set<int> threeBase = { 2, 4, 8, 9 };
+    if (!tupletNumbase) {
+        tupletNumbase = threeBase.count(tupletNum) ? 3 : 2;
+    }
+    // Get number of elements supposed to be in the tuplet _:_:9
+    // Ignore this for the time being
+    tuplet->SetNum(tupletNum);
+    tuplet->SetNumbase(tupletNumbase);
+    m_containerElement = { ElementType::Tuplet, tuplet, tupletNum };
```

Comment:
Woundn't it be necessary to check that we do not have anything else in m_containerElement before assigning this? Or is it done before?  I guess that could happen only with nested tuplets, but it seems to me that in such a case `tuplet` would be lost and we would have a memory leak, wouldn't we?

Reply:
Yes, thanks for pointing this out. We definitely want to clean up previous data if we reach this point without element ever being assigned. Regarding nested `tuplet`, now, that's a different question, since I'm not even sure how such encoding would be achieved in ABC format if supported.